[Ella-e] iP

Ella-e opened the pull request:
# Plaudern 🤖 
> Friendly chatbot that can help to manage your tasks.
- Text-based
- Super fast
- ~~_Easy_~~ SUPER EASY to learn
- **FREE** to use!

## How to use
1. Download the latest version from [here](https://github.com/Ella-e/ip/releases).
2. Double click to install.
3. You are already good to go! :shipit:

## Our progress
- [x] Basic management of tasks (Todo, Deadline, Event).
- [x] Read from and record into local file.
- [ ] More interactive GUI.

## Other purposes
### Can also be used as a source for learning JAVA OOP!
Here's the `main` method:
```java
public static void main(String[] args) {
    Duke duke = new Duke("./data/duke.txt");
    duke.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Ella-e committed:
feat: Level-0: Rename, Greet, Exit

Ella-e committed:
feat: Level 1. Echo

Ella-e committed:
feat: Level 2. Add, List

Ella-e committed:
feat: Level 3. Mark as Done

Ella-e committed:
feat: Level 4. ToDos, Events, Deadlines

Ella-e committed:
feat: A-TextUiTesting: Test using the I/O redirection technique

Ella-e committed:
feat: Level 5. Handle Errors - check whether user input a valid instruction; check whether the instruction's body is emptyl; check if the format for the instruction is correct; check if the mark instruction provides a valid number that is within the availble task list range.

Ella-e committed:
feat: Level 6. Delete - delete tasks from list and error handling.

Ella-e committed:
feat: A-Enums - Adopt enumerates for instruction keywords.

Ella-e committed:
feat: level-7: save into file

Ella-e committed:
feat: Level-8 Store LocalDate object as the time for Deadline class. display the date object in a different form.

Ella-e committed:
feat: refract the project to make it mroe OOP

Ella-e committed:
feat: put code into packages

Ella-e committed:
Merge branch 'add-gradle-support'
A-Gradle

Ella-e committed:
feat: add graddle support

Ella-e committed:
fix: change start directory when exporting jar

Ella-e committed:
feat: add java doc.

Ella-e committed:
feat: better coding standard.

Ella-e committed:
feat: add find functionality

Ella-e committed:
roll back master

Ella-e committed:
Merge branch 'branch-A-CodingStandard'
merge coding standard into master.

Ella-e committed:
master update

Ella-e committed:
merge level-9 into master

Ella-e committed:
fix: handle corrupted file; handle index less than 0 error.

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+    public Duke(String filePath) {
+        this.ui = new UI();
+        this.storage = new Storage(filePath);
+        try {
```
Should this be separated with a newline?

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+				// check for end the session
+				if (parser.getCurrentKey().equals(KeyEnum.EXITKEY)) {
+					break;
+				}
```
Should these lines have this many indentations?

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+				}
+                // continue for the functionality
+                switch (parser.getCurrentKey()) {
+                    case DEADLINE:
```
Should the case clauses have indentations?

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+				if (parser.getCurrentKey().equals(KeyEnum.EXITKEY)) {
+					break;
+				}
+                // continue for the functionality
```
Should this be separated with a newline?

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+                // continue for the functionality
+                switch (parser.getCurrentKey()) {
+                    case DEADLINE:
+                    case TODO:
```
Should the fallthrough cases have comments?

bryanyee33 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+
+	public static void main(String[] args) {
+		Duke duke = new Duke("./data/duke.txt");
+		duke.run();
```
Should these lines have this many indentations?

bryanyee33 commented on src/main/java/duke/exceptions/BaseException.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Base exception for this project.
+ */
+public class BaseException extends RuntimeException{
```
Should there be no space between '{'? I noticed the same issue in several other places too.

bryanyee33 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,116 @@
+ */
+public class Storage {
+	private String filePath = "./data/duke.txt";
+	private File file;
```
Should these lines have this many indentations? I noticed the same issue in several other places too.

bryanyee33 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,116 @@
+            String from = strSplit[3];
+            String to = strSplit[4];
+            task = new Event(status, detail, from, to);
+        break;
```
Should this break have no many indentation?

bryanyee33 commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,52 @@
+     * @return Formatted string to be stored in file.
+     */
+    public String inFileStringFormat() {
+        String intStatus = this.status? "1":"0";
```
Should the ternary expression have no spaces?

bryanyee33 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+package duke.tasks;
+
+import duke.exceptions.*;
```
Should you import *?

bryanyee33 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    public void listTask() {
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println((i+1)+". " + tasks.get(i));
```
Should there be no spaces between '+'?

bryanyee33 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     * @throws EmptyBodyException If the command body is empty.
+     * @throws InvalidDateTimeException If the due date for deadline is not in recognisable format.
+     */
+    public Task addTask(String detail, String from, String to, KeyEnum currentKey) throws EmptyBodyException, InvalidDateTimeException {
```
Should this line be split into 2?

bryanyee33 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+                break;
+            case EVENT:
+                task = new Event(false, detail, from, to);
+                break;
```
Should the case clauses have indentations?

bryanyee33 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+                break;
+        }
+        // Throw empty body exception if the added
+        if (task==null || detail.length() == 0) {
```
Should there be whitespaces? I noticed the same issue in several other places too.

bryanyee33 commented on src/main/java/duke/utils/KeyEnum.java:
```diff
@@ -0,0 +1,22 @@
+ * Enumerate class represents command keywords.
+ */
+public enum KeyEnum {
+    EXITKEY("bye"), LIST("list"), MARK("mark"), UNMARK("unmark"), EVENT("event"), DEADLINE("deadline"), TODO("todo"), DELETE("delete"),INVALID("invalid"), FIND("find");
```
Should this be split into 2 lines? I noticed the same issue in several other places too.

bryanyee33 replied:
Nice use of values within enums 👍

bryanyee33 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,149 @@
+                    throw new WrongFormatException("\"deadline content /from time /to time\"");
+                }
+                break;
+            case MARK:
+            case UNMARK:
```
Should the fallthrough cases have comments?

bryanyee33 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,149 @@
+        String[] userInputSplit = userInput.split(" ");
+        this.determineCurrentKey(userInputSplit[0]);
+        switch (this.currentKey) {
+            case DEADLINE:
```
Should the case clauses have indentations? I noticed the same issue in several other places too.

bryanyee33 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,149 @@
+                } catch (Exception e) {
+                    throw new InvalidNumberException();
+                }
+		}
```
Should this have an indentation?

bryanyee33 reviewed (commented):
Overall, there are many inconsistencies with indentations, spaces within a statement, etc. I would suggest the CheckStyle Gradle once you attempt this week's iP increments.

Ella-e committed:
Fix coding style base on the given standand.

Ella-e committed:
Add checkstyle source files

Ella-e committed:
Merge branch 'branch-A-CheckStyle'
Merge checkstyle

Ella-e committed:
Add GUI

Ella-e committed:
Commit all files

Ella-e committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10

Greg-Tan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,137 @@
+        String response = "";
+        do {
+            response = getResponse(ui.getUserInput());
+        } while (!Objects.equals(response, "Bye. Hope to see you again soon!"));
```
consider defining this string as a constant to avoid magic strings

Greg-Tan commented on src/main/java/duke/control/MainWindow.java:
```diff
@@ -0,0 +1,53 @@
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/user.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/bot.png"));
```
naming convention used looks correct according to the code quality guidelines

Greg-Tan commented on src/main/java/duke/exceptions/EmptyBodyException.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Exception thrown when the command body is empty.
+ */
+public class EmptyBodyException extends BaseException {
```
name of the exception could be more descriptive here. instead of using "empty" could use "missing" to show that body cannot be empty. other exceptions follow this vocabulary by using words such as "invalid" or "illegal"

Greg-Tan commented on src/main/java/duke/ui/UI.java:
```diff
@@ -0,0 +1,115 @@
+     * @param task  Task added.
+     * @param count Number of tasks in the list after the addition.
+     */
+    public String onAddSuccess(Task task, Integer count) {
```
the "success" word may not be necessary since there is already exception handling in your code that produces the error messages. hence, there is no onAddFailure() function to print the error message. I think the function name can also be more descriptive. e.g showAddedTask()

Greg-Tan reviewed (commented):
LGTM. In general, code looks very well done except for a few minor code quality issues with the naming

Ella-e committed:
Add assertions
There are some assumptions in the code.

If they are not satisfied, there might be runtime error.

Let's,
* add assertions to make sure the parser assigns a value to the currentKey.
* add assertions to make sure tasks are added successfully to the tasklist and successfully marked.
* add assertions to make sure tasks to be reformated are not empty.

Assertions in those places check the mistakes made by the programmer and make the code more robust.

Ella-e committed:
Check code quality
The current code has not been checked for its code quality.

Some practices might not be best and makes others hard to understand.

Let's,
* Use checkStyle to detect obvious coding standard mistakes.
* Go through the code manually to see where could be made better.

It helps others to understand the code and also useful for ourselves
to debug.

Ella-e committed:
setup CI

Ella-e committed:
Add reminder of due tasks
There is no reminder of tasks that are due soon.

People might forget their most recent due tasks if the task list contains a lot of tasks.

Let's,
* Show the most recent due tasks when the user enters the app.

It will make the app more useful and userfiendly.

Ella-e committed:
Merge pull request #1 from Ella-e/branch-A-Assertions

Add assertions

Ella-e committed:
Merge branch 'master' into branch-A-CodeQuality
Update code quality

Ella-e committed:
Merge branch 'master' into branch-A-CI
Update with master

Ella-e committed:
Merge branch 'master' into branch-B-Reminders
Update with master.

Ella-e committed:
Merge pull request #2 from Ella-e/branch-A-CodeQuality

Merge Branch A code quality

Ella-e committed:
Merge branch 'master' into branch-A-CI
Update with master

Ella-e committed:
Merge branch 'master' into branch-B-Reminders
Update with master

Ella-e committed:
Merge pull request #3 from Ella-e/branch-A-CI

Branch a ci

Ella-e committed:
Merge branch 'master' into branch-B-Reminders
Update with master

Ella-e committed:
Merge pull request #4 from Ella-e/branch-B-Reminders

Branch b reminders

Ella-e committed:
Add more error handling and more testing

Ella-e committed:
Merge branch 'branch-A-MoreErrorHandling'
The current chatbot will fail to parse the task if the user store the special character | into the
database. Therefore, we should prevent this from happening.

Let's prevent user from using | in their input and add some test cases for the new method we
add.

Ella-e committed:
Make the GUI better.

Ella-e committed:
Merge branch 'branch-A-BetterGui'
Let's make the GUI better.

Ella-e committed:
Reformat test cases

Ella-e committed:
Add more test cases for parser

Ella-e committed:
Fix existing bugs for calculating most recent task.

Ella-e committed:
Add user guide

There is no clear user guide for the project.

Let's:
* Add a clear user guide for the project.
* Add screenshot of product.

Ella-e committed:
Merge branch 'branch-A-UserGuide'

Let's,
* Update user guide
* Update product screenshot
* Fix code quality issues

Ella-e committed:
Update user guide

Ella-e committed:
Update UserGuide

Ella-e committed:
Update user guide

Ella-e committed:
Update test cases

Ella-e committed:
Update user guide to be more user friendly

Ella-e committed:
Add line breaks to user guide

Ella-e committed:
Update UserGuide

Ella-e committed:
Delete lib directory

Ella-e committed:
Delete test_data directory

Ella-e committed:
Add handle end date before start date

Ella-e committed:
Merge branch 'master' of https://github.com/Ella-e/ip

Update the master branch.

Ella-e committed:
Merge branch 'master' of https://github.com/Ella-e/ip

Ella-e committed:
Merge branch 'master' of https://github.com/Ella-e/ip

Ella-e committed:
Update user guide to be more interesting

Ella-e committed:
Remove emoji from user guide

Ella-e committed:
Add warning in user guide

Ella-e committed:
Add list for warning message

Ella-e committed:
Fix long method

Ella-e committed:
Update readme and coding standard

Ella-e committed:
Update readme

Ella-e committed:
Reduce nested code

Ella-e committed:
Reduce nested if statement
